datafusion/physical-expr-common/src/group_value_row.rs

```diff
@@ -0,0 +1,393 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use arrow::array::BooleanBufferBuilder;
+use arrow::array::BufferBuilder;
+use arrow::array::GenericBinaryArray;
+use arrow::array::GenericStringArray;
+use arrow::array::OffsetSizeTrait;
+use arrow::array::PrimitiveArray;
+use arrow::array::{Array, ArrayRef, ArrowPrimitiveType, AsArray};
+use arrow::buffer::NullBuffer;
+use arrow::buffer::OffsetBuffer;
+use arrow::buffer::ScalarBuffer;
+use arrow::datatypes::ArrowNativeType;
+use arrow::datatypes::ByteArrayType;
+use arrow::datatypes::DataType;
+use arrow::datatypes::GenericBinaryType;
+use arrow::datatypes::GenericStringType;
+
+use std::sync::Arc;
+use std::vec;
+
+use crate::binary_map::OutputType;
+use crate::binary_map::INITIAL_BUFFER_CAPACITY;
+
+/// Trait for group values column-wise row comparison
+pub trait ArrayRowEq: Send + Sync {
+    fn equal_to(&self, lhs_row: usize, array: &ArrayRef, rhs_row: usize) -> bool;
+    fn append_val(&mut self, array: &ArrayRef, row: usize);
+    fn len(&self) -> usize;
+    fn is_empty(&self) -> bool;
+    fn build(self: Box<Self>) -> ArrayRef;
+    fn take_n(&mut self, n: usize) -> ArrayRef;
+}
+
+pub struct PrimitiveGroupValueBuilder<T: ArrowPrimitiveType> {
```

Comment:
Actually, when @wiedld and I were discussing this PR, we realized this is quite similar to 

https://github.com/apache/datafusion/blob/f2159e6cae658a0a3f561ec2d15ea948213fd0f8/datafusion/physical-plan/src/aggregates/group_values/primitive.rs#L92

I wonder if we could potentially use that (i.e. impl `ArrayRowEq` for `GroupValuesPrimitive`) 🤔  -- that code is already pretty well tested and optimized 🤔 

Though maybe null handling has to be different

Reply:
There is another difference. `GroupValuesRowLike` can handle mixed type like i32 + utf8. `GroupValuesPrimitive` could only be extended to multiple primitive types but not mixed with utf8. Otherwise, it will be the same as `GroupValuesRowLike`

Reply:
I guess I was imagining that `GroupValuesRowLike` would continue to have a `Vec<Box<dyn ArrayRowEq>>` but we could implement `ArrayRowEq` for GroupValuesPrimtive and GroupValuesBinary, etc

Reply:
Another way is to handle single column with `GroupValuesColumn` (`GroupValuesRowLike`) without slowdown